tests/integration_tests/style/test_rust.py

```diff
@@ -21,40 +24,47 @@ def test_rust_style():
     assert "Diff in" not in stdout
 
 
-def test_ensure_mod_tests():
+def test_ensure_mod_tests(monkeypatch):
     """
     Check that files containing unit tests have a 'tests' module defined.
 
     @type: style
     """
-    # List all source files containing rust #[test] attribute,
-    # (excluding generated files and integration test directories).
-    # Take the list and check each file contains 'mod tests {', output file
-    # name if it doesn't.
-    cmd = (
-        "/bin/bash "
-        "-c "
-        '"grep '
-        "--files-without-match "
-        "'mod tests {' "
-        "\\$(grep "
-        "--files-with-matches "
-        "--recursive "
-        "--exclude-dir=src/*_gen/* "
-        "'\\#\\[test\\]' ../src/*/src)\" "
-        '| grep -v "../src/io_uring/src/bindings.rs"'
-    )
-
-    # The outer grep returns 0 even if it finds files without the match, so we
-    # ignore the return code.
-    result = utils.run_cmd(cmd, no_shell=False, ignore_return_code=True)
-
-    stdout = result.stdout.strip()
-
-    error_msg = (
-        f'Tests found in files without a "tests" module:\n {stdout} '
-        "To ensure code coverage is reported correctly, please check that "
-        'your tests are in a module named "tests".'
-    )
-
-    assert not stdout, error_msg
+    excluding = [
+        "_gen/",
+        "/tests/",
+        "/test_utils",
+        "build/",
+        "src/io_uring/src/bindings.rs",
+    ]
+
+    def cmd(string):
+        cmd = shlex.split(string)
+        process = subprocess.run(cmd, capture_output=True, check=True)
+        return process.stdout.decode("utf-8").split("\n")
+
+    def disjoint(one, two):
+        return list(set(one) - set(two))
+
+    # Re-direct to repository root.
+    monkeypatch.chdir("..")
+
+    # Install ripgrep
+    subprocess.run(["cargo", "install", "ripgrep"], capture_output=True, check=True)
+
+    # Files with `mod tests`.
+    mods = cmd('rg -l "mod tests" -g "*.rs"')
+
+    # Files with `#[test]`.
+    tests = cmd('rg -l "#\\[test\\]" -g "*.rs"')
+
+    # Files with `#[test]` without `mod tests`.
+    tests_without_mods = disjoint(tests, mods)
+
```

Comment:
This is fair but I think the example you have given here is more complex, there are more arguments and tools.

Reply:
Is this even what we want? I thought the goal of this test was to ensure that all tests are inside a `mod test` (unless they're integration tests or generated by bindgen.rs`), but with these regex check a `.rs` file with the structure below passes CI, even though it violates the spirit of the test (although the documentation only says "ensure all modules with tests have a `mod test`", which doesn't rule this case as "bad")

```rs
#[test] 
fn bad() {}

#[cfg(test)]
mod tests {}
```

Reply:
@roypat This issue is present in our current implementation.

Reply:
Yep, I know. I just felt that if we're redoing this test anyway, and we agree that this is not the intended behavior, we might as well fix it.

Reply:
If we want to do a more accurate test, I think we should let go of shelling out to either tool and do the check in Python. I.e sthg like (untested)

```python
for file in Path("..").rglob("*.rs):
    match = re.match(r"#\[test\] .*^mod tests {", file.read_text(), re.MULTILINE)
    assert match is None
```

Reply:
Thinking about it some more, I think any attempt at solving this using regex breaks on 

```rs
#[cfg(test)]
mod test {}

#[test]
fn bad() {}
```

because a regex cannot know when the `mod test` is over - bracket matching is the sterotypical context free language that is not regular

Reply:
I think a fully accurate solution would ultimately need to consider the Rust AST (akin to https://docs.rs/syn/latest/syn/struct.File.html). My overall view is this test is not worth the effort, in the worst case a test gets included in coverage and raises coverage slightly higher than it should have, I don't think this is a significant impact.

Reply:
yeah, I agree on this.

Reply:
We came to the conclusion to merge this PR.